libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp

```diff
@@ -0,0 +1,453 @@
+      if (j >= odfAutoCorr.size())
+         break;
+      while (true)
+      {
+         const auto i = MapToPositiveHalfIndex(j - 1, odfAutocorrFullSize);
+         const auto k = MapToPositiveHalfIndex(j + 1, odfAutocorrFullSize);
+         if (
+            odfAutoCorr[i] <= odfAutoCorr[j] &&
+            odfAutoCorr[j] >= odfAutoCorr[k])
+            break;
+         j = odfAutoCorr[i] > odfAutoCorr[k] ? i : k;
+      }
+      sum += odfAutoCorr[j];
+      ++numIndices;
+      if (debugOutput)
+         debugOutput->odfAutoCorrPeakIndices.push_back(j);
+   }
+   return sum / numIndices;
+}
+
+size_t GetBestBarDivisionIndex(
+   const std::vector<BarDivision>& possibleBarDivisions,
+   double audioFileDuration, int numTatums, const std::vector<float>& odf,
+   QuantizationFitDebugOutput* debugOutput)
+
+{
+   const auto odfAutoCorr = GetNormalizedCircularAutocorr(odf);
+   if (debugOutput)
+      debugOutput->odfAutoCorr = odfAutoCorr;
+   const auto odfAutocorrFullSize = 2 * (odfAutoCorr.size() - 1);
+   assert(IsPowOfTwo(odfAutocorrFullSize));
+   const auto odfAutoCorrSampleRate = odfAutocorrFullSize / audioFileDuration;
+
+   std::vector<double> scores(possibleBarDivisions.size());
+   // These (still) only depend on the beat rate. We look at
+   // self-similarities at beat intervals, so to say, without examining the
+   // contents of the beats. Since several bar divisions may yield the same
+   // number of beats, we may be able to re-use some calculations.
+   std::unordered_map<int /*numBeats*/, double> autocorrScoreCache;
+   std::transform(
+      possibleBarDivisions.begin(), possibleBarDivisions.end(), scores.begin(),
+      [&](const BarDivision& barDivision) {
+         const auto numBeats = barDivision.numBars * barDivision.beatsPerBar;
+         const auto timeSignature = GetTimeSignature(barDivision, numTatums);
+         const auto bpm = 1. * numBeats / audioFileDuration * 60;
+         const auto likelihood = GetTimeSignatureLikelihood(timeSignature, bpm);
+         if (!autocorrScoreCache.count(numBeats))
+            autocorrScoreCache[numBeats] = GetBeatSelfSimilarityScore(
+               odfAutoCorrSampleRate, bpm, odfAutoCorr, odfAutocorrFullSize,
+               debugOutput);
+         const auto selfSimilarityScore = autocorrScoreCache.at(numBeats);
+         return likelihood * selfSimilarityScore;
+      });
+
+   return std::max_element(scores.begin(), scores.end()) - scores.begin();
+}
+
+MusicalMeter GetMostLikelyMeterFromQuantizationExperiment(
+   const std::vector<float>& odf, int numTatums,
+   const std::vector<BarDivision>& possibleBarDivisions,
+   double audioFileDuration, QuantizationFitDebugOutput* debugOutput)
+{
+   const auto fourFourIt = std::find_if(
+      possibleBarDivisions.begin(), possibleBarDivisions.end(),
+      [&](const BarDivision& barDivision) {
+         return GetTimeSignature(barDivision, numTatums) ==
+                TimeSignature::FourFour;
```

Comment:
Could there be more than one hypthetical bar division with a four/four signature?  If so, why should the first be preferred?

Reply:
Absolutely, thanks.